### PR TITLE
feat(element-templates) add support for field injections Closes #483

### PR DESCRIPTION
**Pull request by felix-mueller**
Fixes #483
https://github.com/camunda/camunda-modeler/issues/483

### Proposed Changes

* Make it possible to add fieldinjections to element-templates (string and expressions)

### P.S. This is my first PR on bpmn-js, so handle with care 💃 💣

### CODE REVIEW

**Comment by philippfromme** [comment]
@nikku Who should do the review?

**Comment by nikku** [comment]
Do you want to have a look? 🍨

**Comment by philippfromme** [comment]
I can, but it will take me while 🤣

File: lib/provider/camunda/element-templates/CreateHelper.js [hunk 1]
```diff
@@ -174,6 +176,35 @@ function createCamundaExecutionListenerScript(binding, value, bpmnFactory) {
...
+ *
+ * @return {ModdleElement}
+ */
+function createCamundaFieldInjection(binding, value, bpmnFactory) {
+  var DEFAULT_PROPS = {
+    'string': undefined,
+    'expression': undefined,
+    'name': undefined
+  };
+  var CAMUNDA_FIELD_EXTENSION_ELEMENT = 'camunda:Field';
```
Comment (nikku) [inline lib/provider/camunda/element-templates/CreateHelper.js]: Please inline the variables directly and don't CAPSLOCK them. We're generally only using ALL_CAPS, if it is a global constant.

File: lib/provider/camunda/element-templates/CreateHelper.js [hunk 2]
```diff
@@ -174,6 +176,35 @@ function createCamundaExecutionListenerScript(binding, value, bpmnFactory) {
...
+  var props = assign({}, DEFAULT_PROPS);
+
+  if (!binding.expression) {
+    props.string = value.value;
+  } else {
+    props.expression = value.value;
+  }
+  props.name = binding.name;
+
+  return bpmnFactory.create(CAMUNDA_FIELD_EXTENSION_ELEMENT, props);
```
Comment (nikku) [inline lib/provider/camunda/element-templates/CreateHelper.js]: Think about simply inlining the `camunda:field` value here.

File: lib/provider/camunda/element-templates/CreateHelper.js (hunk 1 above)
Comment (nikku) [reply lib/provider/camunda/element-templates/CreateHelper.js]: Please rename `camunda:Field` to `camunda:field`. This is in line with the names of other extension points (`camunda:executionListener`, ...).

File: lib/provider/camunda/element-templates/parts/CustomProps.js [hunk 3]
```diff
@@ -609,6 +624,20 @@ function setPropertyValue(element, property, value, bpmnFactory) {
...
+  if (bindingType === CAMUNDA_FIELD) {
+    var existingFieldInjections = findExtensions(bo, [ 'camunda:Field' ]);
+    var newFieldInjections = createCamundaFieldInjection(binding, value, bpmnFactory);
+    updates.push(cmdHelper.addAndRemoveElementsFromList(
+      element,
+      extensionElements,
+      'values',
+      null,
+      [ newFieldInjections ],
+      existingFieldInjections ? [ existingFieldInjections ] : []
```
Comment (nikku) [inline lib/provider/camunda/element-templates/parts/CustomProps.js]: Shouldn't this be `existingFieldInjections ? [ existingFieldInjections ] : []`? There is definitely a test case missing here.

File: test/spec/provider/camunda/element-templates/ValidatorSpec.js [hunk 4]
```diff
@@ -205,7 +205,7 @@ describe('element-templates - Validator', function() {
       'invalid property type <InvalidType>; must be any of { String, Text, Boolean, Hidden, Dropdown }',
       'invalid property.binding type <alsoInvalid>; must be any of { ' +
         'property, camunda:property, camunda:inputParameter, ' +
-        'camunda:outputParameter, camunda:in, camunda:out, camunda:in:businessKey, camunda:executionListener }'
+        'camunda:outputParameter, camunda:in, camunda:out, camunda:in:businessKey, camunda:executionListener, camunda:Field }'
```
Comment (nikku) [inline test/spec/provider/camunda/element-templates/ValidatorSpec.js]: Split up into new line, please.

File: test/spec/provider/camunda/element-templates/ValidatorSpec.js [hunk 5]
```diff
@@ -339,7 +339,7 @@ describe('element-templates - Validator', function() {
...
```
Comment (nikku) [inline test/spec/provider/camunda/element-templates/ValidatorSpec.js]: Split up into new line, please.

File: test/spec/provider/camunda/element-templates/cmd/ChangeElementTemplateHandlerSpec.js [hunk 6]
```diff
@@ -787,6 +787,91 @@ describe('element-templates - cmd', function() {
 
     });
```
Comment (nikku) [inline test/spec/provider/camunda/element-templates/cmd/ChangeElementTemplateHandlerSpec.js]: Two lines between specs / describes.

File: test/spec/provider/camunda/element-templates/cmd/fieldinjections.json [hunk 7]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "Valid",
+  "id": "foo",
+  "appliesTo": [
+    "bpmn:ServiceTask"
+  ],
+  "properties": [
+  {
```
Comment (nikku) [inline test/spec/provider/camunda/element-templates/cmd/fieldinjections.json]: Please indent properly:

```json
"properties": [
  {
    ...
  },
  {
    ...
  }
], ...
```

File: test/spec/provider/camunda/element-templates/cmd/ChangeElementTemplateHandlerSpec.js [hunk 8]
```diff
@@ -787,6 +787,91 @@ describe('element-templates - cmd', function() {
...
+
+        var fieldInjections = findExtensions(taskShape, [ 'camunda:Field' ]);
+
+        // then
+        expect(task.get('camunda:modelerTemplate')).to.eql(currentTemplate.id);
+
+        expect(fieldInjections).to.exist;
+      }));
+
+    });
```
Comment (nikku) [inline test/spec/provider/camunda/element-templates/cmd/ChangeElementTemplateHandlerSpec.js]: Add test case that verify:

* camunda:Field on Connectors
* interaction with existing Camunda Field declarations (will they be kept or deleted)?
* camunda field updating

**Comment by felix-mueller** [comment]
Hi @nikku 
thanks for your feedback.
I added most proposed test cases. I am just wondering about "camunda:Field on Connectors"
Why is this test necessary? Is there anything special about connectors for field injections?
Can you give me more details what kind of test case you would expect here?

- [ ] camunda:Field on Connectors
- [x] interaction with existing Camunda Field declarations (will they be kept or deleted)?
- [x] camunda field updating

I hope my other updates are fine.

**Comment by nikku** [comment]
> Is there anything special about connectors for field injections?

I'm not sure right now, whether or not field injections should work on connectors, too (I guess not)? If they should work however, we'd need to have test cases for them.

**Comment by nikku** [comment]
The other updates look very promising. 👍

**Comment by nikku** [comment]
Looks like field injections are not supported on connectors but [on certain event definitions, task and execution listeners only](https://docs.camunda.org/manual/7.6/user-guide/process-engine/delegation-code/#field-injection). 

These are out of scope for this ticket. So looks like from your

**Comment by nikku** [comment]
Merged via https://github.com/bpmn-io/bpmn-js-properties-panel/commit/a057edf5c0addbfbfa05f4962d45c3d4acb72200.

Congratulations to your first, quite complex PR being merged 🥁.

**Comment by felix-mueller** [comment]
Thanks @nikku 🙌 😃

**Pull request closed without merging**